common/context/src/main/java/io/helidon/common/context/ListContext.java

```diff
@@ -0,0 +1,221 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.context;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * A {@link Context} implementation with deque registry.
+ */
+class ListContext implements Context {
+
+    private final Context parent;
+    private final ConcurrentHashMap<Object, ClassifiedRegistry> classifiers = new ConcurrentHashMap<>();
+    private final ClassifiedRegistry registry = new ClassifiedRegistry();
+
+    /**
+     * Creates new instance with defined parent.
+     *
+     * @param parent a parent context or {@code null}.
+     */
+    ListContext(Context parent) {
+        this.parent = parent;
+    }
+
+    /**
+     * Creates new instance.
+     */
+    ListContext() {
+        this(null);
+    }
+
+    @Override
+    public <T> void register(T instance) {
+        registry.register(instance);
+    }
+
+    @Override
+    public <T> void supply(Class<T> type, Supplier<T> supplier) {
+        registry.supply(type, supplier);
+    }
+
+    @Override
+    public <T> Optional<T> get(Class<T> type) {
+        T result = registry.get(type);
+        if (result == null) {
+            if (parent == null) {
+                return Optional.empty();
+            } else {
+                return parent.get(type);
+            }
+        } else {
+            return Optional.of(result);
+        }
+    }
+
+    @Override
+    public <T> void register(Object classifier, T instance) {
+        Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
+        ClassifiedRegistry cr = classifiers.computeIfAbsent(classifier, k -> new ClassifiedRegistry());
+        cr.register(instance);
+    }
+
+    @Override
+    public <T> void supply(Object classifier, Class<T> type, Supplier<T> supplier) {
+        Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
+        ClassifiedRegistry cr = classifiers.computeIfAbsent(classifier, k -> new ClassifiedRegistry());
+        cr.supply(type, supplier);
+    }
+
+    @Override
+    public <T> Optional<T> get(Object classifier, Class<T> type) {
+        Objects.requireNonNull(classifier, "Parameter 'classifier' is null!");
+        ClassifiedRegistry cr = classifiers.get(classifier);
+        if (cr != null) {
+            T result = cr.get(type);
+            if ((result == null) && (parent != null)) {
```

Comment:
These internal parens are not required, but... I'm assuming you find this clearer? If so, I'm pretty sure that this is not consistent across the project.

Reply:
My setting of IDE - "add clarifying parentheses" - I do prefer to clearly separate the conditions.
We should add this to the list of known inconsistencies and eventually discuss and agree on a single approach.
BTW I am doing this since the beginning of the project...

Reply:
Creating a follow up issue for this: #648